packages/core/src/utils/cache.ts

```diff
@@ -0,0 +1,24 @@
+import { FunctionType } from './types';
+
+const createCache = <TFunction extends FunctionType>(
+  func: TFunction,
+  comparer: (prev: Parameters<TFunction>, next: Parameters<TFunction>) => boolean,
+): (...arg: Parameters<TFunction>) => ReturnType<TFunction> => {
+  let cachedArg: Parameters<TFunction>;
+  let cachedResult: ReturnType<TFunction>;
+
+  return (...arg: Parameters<TFunction>) => {
+    const comparerResult = cachedArg !== undefined && comparer(cachedArg, arg);
+
+    if (comparerResult) {
+      return cachedResult;
+    }
+
+    cachedArg = arg;
+    cachedResult = func(...arg) as ReturnType<TFunction>;
+
+    return cachedResult;
+  };
```

Comment:
Not sure my solution is better but take a look please
```suggestion
  let cachedArg: Parameters<TFunction>;
  let cachedResult: ReturnType<TFunction>;

  const callAndCache =  (...arg: Parameters<TFunction>) => {
    cachedArg = arg;
    return cachedResult = func(...arg) as ReturnType<TFunction>;
  };

  const callCached = (...arg: Parameters<TFunction>) => {
    return comparer(cachedArg, arg)
        ? cachedResult
        : callAndCache(...arg) as ReturnType<TFunction>;
  };

  let result = (...arg: Parameters<TFunction>) => {
    result = callCached;
    return callAndCache(...arg) as ReturnType<TFunction>;
  };

  return (...arg: Parameters<TFunction>) => result(...arg) as ReturnType<TFunction>;
```

Reply:
I look into your solution and agree with you that is a cool technic.
Maybe, it's a little more complicated than mine, but let's try it.

I will push my variant of this idea.